include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -0,0 +1,190 @@
+    scalar_type D(scalar_type a2, scalar_type NdotH2)
+    {
+        scalar_type denom = NdotH2 * (a2 - scalar_type(1.0)) + scalar_type(1.0);
+        return a2 * numbers::inv_pi<scalar_type> / (denom * denom);
+    }
+
+    scalar_type FVNDF(scalar_type fresnel_ndf, scalar_type G1_over_2NdotV, bool transmitted, scalar_type VdotH, scalar_type LdotH, scalar_type VdotHLdotH, scalar_type orientedEta)
+    {
+        scalar_type FNG = fresnel_ndf * G1_over_2NdotV;
+        scalar_type factor = scalar_type(0.5);
+        if (transmitted)
+        {
+            const scalar_type VdotH_etaLdotH = (VdotH + orientedEta * LdotH);
+            // VdotHLdotH is negative under transmission, so this factor is negative
+            factor *= -scalar_type(2.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH);
+        }
+        return FNG * factor;
+    }
+
+    scalar_type DG1(scalar_type ndf, scalar_type G1_over_2NdotV)
+    {
+        return ndf * scalar_type(0.5) * G1_over_2NdotV;
+    }
+
+    scalar_type DG1(scalar_type ndf, scalar_type G1_over_2NdotV, bool transmitted, scalar_type VdotH, scalar_type LdotH, scalar_type VdotHLdotH, scalar_type orientedEta, scalar_type reflectance)
+    {
+        scalar_type FN = hlsl::mix(reflectance, scalar_type(1.0) - reflectance, transmitted) * ndf;
+        return FVNDF(FN, G1_over_2NdotV, transmitted, VdotH, LdotH, VdotHLdotH, orientedEta);
+    }
+
+    scalar_type devsh_part(scalar_type NdotX2, scalar_type a2, scalar_type one_minus_a2)
+    {
+        return sqrt(a2 + one_minus_a2 * NdotX2);
+    }
+
+    scalar_type G1_wo_numerator(scalar_type NdotX, scalar_type NdotX2, scalar_type a2, scalar_type one_minus_a2)
+    {
+        return scalar_type(1.0) / (NdotX + devsh_part(NdotX2,a2,one_minus_a2));
+    }
+
+    scalar_type G1_wo_numerator(scalar_type NdotX, scalar_type devsh_part)
+    {
+        return scalar_type(1.0) / (NdotX + devsh_part);
+    }
+
+    scalar_type correlated_wo_numerator(scalar_type a2, scalar_type NdotV, scalar_type NdotV2, scalar_type NdotL, scalar_type NdotL2)
+    {
+        scalar_type one_minus_a2 = scalar_type(1.0) - a2;
+        scalar_type Vterm = NdotL * devsh_part(NdotV2, a2, one_minus_a2);
+        scalar_type Lterm = NdotV * devsh_part(NdotL2, a2, one_minus_a2);
+        return scalar_type(0.5) / (Vterm + Lterm);
+    }
+
+    scalar_type G2_over_G1(scalar_type a2, bool transmitted, scalar_type NdotV, scalar_type NdotV2, scalar_type NdotL, scalar_type NdotL2)
+    {
+        scalar_type G2_over_G1;
+        if (transmitted)
+        {
+            scalar_type one_minus_a2 = scalar_type(1.0) - a2;
+            scalar_type devsh_v = devsh_part(NdotV2, a2, one_minus_a2);
+            scalar_type L_v = scalar_type(0.5) * (devsh_v / NdotV - scalar_type(1.0));
+            scalar_type L_l = scalar_type(0.5) * (devsh_part(NdotL2, a2, one_minus_a2) / NdotL - scalar_type(1.0));
+            G2_over_G1 = hlsl::beta<scalar_type>(scalar_type(1.0) + L_l, scalar_type(1.0) + L_v);
```

Comment:
if you're gonna add 1 to L_v and L_l, then just compute them right away turn that -1 into a +1

Reply:
uses `bxdf::beta` now that does check only once

Reply:
continued in https://github.com/Devsh-Graphics-Programming/Nabla/pull/916/files/e7fcfa3982a3beb9f064ad8353df6126e6b84ce0..ff830452df22182988e57576a559d42e8c488549#r2306794747